### PR TITLE
HDDS-10112. Dashboard for read key metrics

### PR DESCRIPTION
**Pull request by muskan1012**
## What changes were proposed in this pull request?

Created a new dashboard and added visualisations for all the required ReadKey Metrics.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-10112

## How was this patch tested?

The patch has been tested over a Grafana.
image

### CODE REVIEW

**Review by tanvipenumudy — COMMENTED** [review]
Thank you @muskan1012, for working on the patch. 

Could you please share snapshots to the added dashboards over the PR?
Here are a few references: [Dashboard for Chunks & Volume IO](https://github.com/apache/ozone/pull/6554), [S3 API-level Dashboard](https://github.com/apache/ozone/pull/6202)

**Comment by tanvipenumudy** [comment]
@devabhishekpal could you also please take a look? Thanks!

**Comment by muskan1012** [comment]
Sure @tanvipenumudy, please find snapshots here: 
image
image
image

**Review by devabhishekpal — APPROVED** [review]
This looks good to me.
Thanks @muskan1012 for the change.
+1

File: hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json [hunk 1]
```diff
@@ -0,0 +1,3664 @@
...
+            }
+          },
+          "targets": [
+            {
+              "datasource": {
+                "type": "prometheus"
+              },
+              "disableTextWrap": false,
+              "editorMode": "code",
+              "expr": "rdb_metrics_num_db_key_get_if_exist_checks{instance=~\"ccycloud-.*.mus-t1.root.comops.site:9875\"}",
```
Comment (tanvipenumudy) [inline hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json:479]: Please fix the hardcoded host value here.

File: hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json [hunk 2]
```diff
@@ -0,0 +1,3664 @@
...
+              "refId": "A",
+              "useBackend": false
+            },
+            {
+              "datasource": {
+                "type": "prometheus"
+              },
+              "disableTextWrap": false,
+              "editorMode": "code",
+              "expr": "rdb_metrics_num_db_key_get_if_exist_gets{instance=~\"ccycloud-.*.mus-t1.root.comops.site:9875\"}",
```
Comment (tanvipenumudy) [inline hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json:494]: Same as the previous comment, please fix the hardcoded host value here.

File: hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json [hunk 3]
```diff
@@ -0,0 +1,3664 @@
...
+            }
+          },
+          "targets": [
+            {
+              "datasource": {
+                "type": "prometheus"
+              },
+              "disableTextWrap": false,
+              "editorMode": "builder",
+              "expr": "table_cache_metrics_hit_count{instance=~\"ccycloud-.*.mus-t1.root.comops.site:9875\"}",
```
Comment (tanvipenumudy) [inline hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json:1613]: Same as the previous comment, please fix the hardcoded host value here as well.

File: hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json [hunk 4]
```diff
@@ -0,0 +1,3664 @@
...
+              "refId": "A",
+              "useBackend": false
+            },
+            {
+              "datasource": {
+                "type": "prometheus"
+              },
+              "disableTextWrap": false,
+              "editorMode": "code",
+              "expr": "table_cache_metrics_miss_count{instance=~\"ccycloud-.*.mus-t1.root.comops.site:9875\"}",
```
Comment (tanvipenumudy) [inline hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json:1628]: Same as the previous comments.

File: hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json [hunk 5]
```diff
@@ -0,0 +1,3664 @@
...
+            }
+          },
+          "targets": [
+            {
+              "datasource": {
+                "type": "prometheus"
+              },
+              "disableTextWrap": false,
+              "editorMode": "code",
+              "expr": "ugi_metrics_login_success_avg_time{instance=~\"ccycloud-.*.mus-t1.root.comops.site:9877\"}",
```
Comment (tanvipenumudy) [inline hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json:3535]: Please review this instance as well.

File: hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json [hunk 6]
```diff
@@ -0,0 +1,3664 @@
...
+            }
+          },
+          "targets": [
+            {
+              "datasource": {
+                "type": "prometheus"
+              },
+              "disableTextWrap": false,
+              "editorMode": "code",
+              "expr": "ugi_metrics_login_success_num_ops{instance=~\"ccycloud-.*.mus-t1.root.comops.site:9877\"}",
```
Comment (tanvipenumudy) [inline hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json:3630]: Same as the previous comments, please review this instance as well.

File: hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json [hunk 7]
```diff
@@ -0,0 +1,3664 @@
...
+              "hide": false,
+              "includeNullMetadata": true,
+              "instant": false,
+              "legendFormat": "{{__name__}}, {{hostname}}",
+              "range": true,
+              "refId": "B",
+              "useBackend": false
+            }
+          ],
+          "title": "Rdb no. of db key metrics",
```
Comment (tanvipenumudy) [inline hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json:505]: We may keep the heading consistent with the other charts under the same section.
```suggestion
          "title": "Rocksdb no. of db key metrics",
```

File: hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json [hunk 8]
```diff
@@ -0,0 +1,3664 @@
...
+              "fullMetaSearch": false,
+              "includeNullMetadata": true,
+              "instant": false,
+              "legendFormat": "{{__name__}}, {{hostname}}",
+              "range": true,
+              "refId": "A",
+              "useBackend": false
+            }
+          ],
+          "title": "Om num key metrics",
```
Comment (tanvipenumudy) [inline hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json:1200]:
```suggestion
          "title": "OM num key metrics",
```

File: hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json [hunk 9]
```diff
@@ -0,0 +1,3664 @@
...
+              "fullMetaSearch": false,
+              "includeNullMetadata": true,
+              "instant": false,
+              "legendFormat": "{{__name__}}, {{hostname}}",
+              "range": true,
+              "refId": "A",
+              "useBackend": false
+            }
+          ],
+          "title": "Om num key ops metrics",
```
Comment (tanvipenumudy) [inline hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json:1295]:
```suggestion
          "title": "OM num key ops metrics",
```

File: hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json [hunk 10]
```diff
@@ -0,0 +1,3664 @@
...
+              "range": true,
+              "refId": "A",
+              "useBackend": false
+            }
+          ],
+          "title": "Get service lists metrics",
+          "type": "timeseries"
+        }
+      ],
+      "title": "OM Servicelists Metrics",
```
Comment (tanvipenumudy) [inline hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json:1408]:
```suggestion
      "title": "OM Service Lists Metrics",
```

File: hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json [hunk 11]
```diff
@@ -0,0 +1,3664 @@
...
+              "range": true,
+              "refId": "A",
+              "useBackend": false
+            }
+          ],
+          "title": "Read key info (avg time) metrics",
+          "type": "timeseries"
+        }
+      ],
+      "title": "OM Read key Info Metrics",
```
Comment (tanvipenumudy) [inline hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json:1518]:
```suggestion
      "title": "OM Read Key Info Metrics",
```

File: hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json [hunk 12]
```diff
@@ -0,0 +1,3664 @@
...
+              "hide": false,
+              "includeNullMetadata": true,
+              "instant": false,
+              "legendFormat": "{{__name__}}, {{hostname}}",
+              "range": true,
+              "refId": "C",
+              "useBackend": false
+            }
+          ],
+          "title": "Rocksdb level0 metrics (num files )",
```
Comment (tanvipenumudy) [inline hadoop-ozone/dist/src/main/compose/common/grafana/dashboards/Ozone - ReadKey Metrics.json:394]:
```suggestion
          "title": "Rocksdb level0 metrics (num files)",
```

**Review by tanvipenumudy — COMMENTED** [review]
Please find a few minor nits.

**Comment by muskan1012** [comment]
@tanvipenumudy thanks for the review, I have addressed all of your comments and fixed minor nits as well.

Here I am attaching the Snapshot of the Dashboard with updated json.
image
image
image

**Comment by muskan1012** [comment]
Thank you @devabhishekpal for reviewing it.

**Review by tanvipenumudy — APPROVED** [review]
Thank you @muskan1012 for working on the patch, the changes LGTM. Thank you @devabhishekpal for reviewing.

**Pull request merged**